server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+                        // snapshots assigned and we have to assign them now based on the new RepositoryData that resulted from the delete.
+                        final SnapshotsInProgress snapshotsInProgress = currentState.custom(SnapshotsInProgress.TYPE);
+                        // There is one special case to deal with here: If the RepositoryData passed to this method is null, then that
+                        // means that the delete operation failed to load the current RepositoryData, indicating an issue with the
+                        // repository so we fail all queued up operations.
+                        final boolean failAllQueuedOperations = repositoryData == null;
+                        if (failAllQueuedOperations) {
+                            deletions.getEntries().stream().filter(entry -> entry.repository().equals(deleteEntry.repository()))
+                                    .map(SnapshotDeletionsInProgress.Entry::uuid).forEach(deletionsToFail::add);
+                            // We failed to read repository data for this delete, it is not the job of SnapshotsService to
+                            // retry this kind of issue so we fail all the pending deletes
+                            deletions = deletions.withRemovedRepository(deleteEntry.repository());
+                        }
+                        final SnapshotsInProgress updatedSnapshotsInProgress;
+                        if (snapshotsInProgress == null) {
+                            updatedSnapshotsInProgress = new SnapshotsInProgress();
+                        } else {
+                            final List<SnapshotsInProgress.Entry> snapshotEntries = new ArrayList<>();
+
+                            // Keep track of shardIds that we started snapshots for as a result of removing this delete so we don't assign
+                            // them to multiple snapshots by accident
+                            final Set<ShardId> reAssignedShardIds = new HashSet<>();
+
+                            for (SnapshotsInProgress.Entry entry : snapshotsInProgress.entries()) {
+                                if (entry.repository().equals(deleteEntry.repository())) {
+                                    if (failAllQueuedOperations) {
+                                        // We failed to read repository data for this delete, it is not the job of SnapshotsService to
+                                        // retry these kinds of issues so we fail all the pending snapshots
+                                      snapshotsToFail.add(entry.snapshot());
+                                    } else if (entry.state().completed() == false) {
+                                        boolean updatedQueuedSnapshot = false;
+                                        for (ObjectCursor<ShardSnapshotStatus> value : entry.shards().values()) {
+                                            if (value.value.equals(ShardSnapshotStatus.UNASSIGNED_WAITING)) {
```

Comment:
As mentioned in Slack, this could possibly be result in shards to be done out-of-snapshot-order in case where there is a relocating primary.

Reply:
Fixed this case now in https://github.com/elastic/elasticsearch/pull/56911/commits/ee8827588c76bba0ab2181691a99027aa6df90be and added test+assertion coverage for it.

:100: thanks for spotting this!